core/src/main/java/org/apache/druid/java/util/common/parsers/TimestampParser.java

```diff
@@ -126,14 +126,24 @@
     return o -> {
       Preconditions.checkNotNull(o, "null timestamp");
 
-      if (o instanceof Number) {
+      if (o instanceof Number && isNumericFormat(format)) {
```

Comment:
`isNumericFormat(format)` does not need to be computed for each object. You can compute this outside the function that is applied per object (on line 125)